FEM/Output.cpp

```diff
@@ -1487,6 +1595,170 @@ void COutput::WriteELEValuesTECHeader(fstream& tec_file)
     tec_file << "\n";
 }
 
+/**************************************************************************
+   FEMLib-Method:
+   Task:
+   Programing:
+   09/2019 JT
+**************************************************************************/
+void COutput::WriteELECellCenteredValuesTECData(fstream& tec_file, int e_type)
+{
+    CRFProcess* m_pcs_2 = NULL;
+    if (_ele_value_vector.empty())
+        return;
+
+    // output of node coordinates (not necesary to output all, but simpler here)
+    if (!tecplot_zone_share || !_new_file_opened)
+    {
+    // streams for buffering since block output is mandatory
+        std::stringstream y_vals, z_vals;
+        y_vals.setf(tec_file.flags());
+        y_vals.precision(tec_file.precision());
+        z_vals.setf(tec_file.flags());
+        z_vals.precision(tec_file.precision());
+        tec_file << "#x Coordinates:\n";
+        y_vals << "#y Coordinates:\n";
+        z_vals << "#z Coordinates:\n";
+        for (size_t j = 0; j < m_msh->GetNodesNumber(false); j++)
+        {
+            CNode* node = m_msh->nod_vector[j];  // 23.01.2013. WW
+            const double* x = node->getData();
+            tec_file << x[0] << " ";
+            y_vals << x[1] << " ";
+            z_vals << x[2] << " ";
+            if (j>0 && j%10==0)
+            {
+                tec_file << "\n";
+                y_vals << "\n";
+                z_vals << "\n";
+            }
+        }
+        tec_file << "\n";
+        y_vals << "\n";
+        z_vals << "\n";
+        tec_file << y_vals.rdbuf() << z_vals.rdbuf();
+
+    }
+
+    vector<bool> skip;  // CB
+    size_t no_ele_values = _ele_value_vector.size();
+    bool out_element_vel = false;
+    bool out_element_transport_flux = false;    // JOD 2014-11-10
+    for (size_t j = 0; j < no_ele_values; j++)  // WW
+    {
+        if (_ele_value_vector[j].find("VELOCITY") != string::npos)
+        {
+            out_element_vel = true;
+            // break;  // CB: allow output of velocity AND other ele values
+            skip.push_back(false);
+        }
+        else if (_ele_value_vector[j].find("TRANSPORT_FLUX") !=
+                 string::npos)  // JOD 2014-11-10
+        {
+            out_element_transport_flux = true;
+            skip.push_back(false);
+        }
+        else
+        {
+            m_pcs_2 = GetPCS_ELE(_ele_value_vector[j]);
+            skip.push_back(true);
+        }
+    }
+    vector<int> ele_value_index_vector(no_ele_values);
+    GetELEValuesIndexVector(ele_value_index_vector);
+
+    // streams for buffering since block output is mandatory
+    //count trues in skip for additional streams
+    int n_additional_streams(0);
+    for (size_t i= 0; i!= skip.size();++i)
+        if (skip[i])
+            ++n_additional_streams;
+
+    std::vector<std::stringstream*> streams(2+n_additional_streams);
```

Comment:
What do you think about this way to output in block format(sorted by quantities instead of by nodes)? I decided to temporarily store the required output in streams in order to just loop once on the elements. Not sure if it is the fastest/ most efficient way ...

Reply:
at the moment, i don’t find any critical problems with it. we can improve the code later if it makes a serious problem.